Porting GPU sphinx docs (#900)
LLNL/axom: src/docs/sphinx/dev_guide/gpu_porting.rst

> .. ## Copyright (c) 2017-2022, Lawrence Livermore National Security,con LLC and

Reviewer: Since this is a porting of the document, not the writing, this comment may not apply.  But as a newbie reading this doc, I have one comment.  This page does a good job of introducing me to what I have to do to port to GPU, but to write code, I'd have to get more details from the header files.  Some references to where the headers and how they are organized into directories would be helpful.

	* ``AXOM_USE_HIP`` - Defined if Axom is built with HIP

* ``AXOM_DEVICE_CODE`` - Denotes code will be compiled on device
* ``AXOM_GPUCC`` - Denotes code will be compiled with HIP or CUDA compiler

For ``__host__ __device__`` decoration::

  //---------------------------------------------------------------------------
  // Functions
  //---------------------------------------------------------------------------

  #define AXOM_DEVICE __device__

  #define AXOM_HOST_DEVICE __host__ __device__

  #define AXOM_HOST __host__

  //---------------------------------------------------------------------------
  // Lambdas
  //---------------------------------------------------------------------------

  #define AXOM_LAMBDA [=] AXOM_HOST_DEVICE

  #define AXOM_DEVICE_LAMBDA [=] AXOM_DEVICE

  #define AXOM_HOST_LAMBDA [=] AXOM_HOST

.. _gpu-memory-label:

===============
Memory 
===============

Link to Axom's `memory management header file`_

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Memory Management Routines
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Umpire has the concept of "allocators" associated with each
`memory resource type`_ (``umpire::resource::MemoryResourceType``)

> To allocate memory on a particular resource, you use the ID associated with the

Reviewer: Suggest rephrasing to make reading less awkward: "you use the ID for the allocator associated with the MemoryResourceType."

.. ## Copyright (c) 2017-2022, Lawrence Livermore National Security,con LLC and
.. ## other Axom Project Developers. See the top-level LICENSE file for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)

.. _gpu-porting-label:

*******************************
GPU Porting in Axom
*******************************

Axom uses RAJA and Umpire as the main workhorses for GPU porting,

> with a set of convenience macros, Axom name-spaced wrappers around commonly

Reviewer 2: Suggestion: Would the following read slightly better?
```suggestion
with a set of convenience macros, `axom`-namespaced wrappers around commonly
```

used functions, and preset execution spaces for host/device execution.

> 

Reviewer 2: It might be good to mention something about our kernel and memory space support when axom is configured without RAJA and Umpire respectively. (It doesn't have to be here though).

Specifically, `axom::forall` works and so do our memory functions, but they only work on the default space (i.e. `host`/CPU) and with the std allocator.

time or for ``__host__ __device__`` decoration of functions/lambdas.

For guarding device code:

* ``AXOM_USE_GPU`` - Defined if Axom is built with CUDA or HIP

	* ``AXOM_USE_CUDA`` - Defined if Axom is built with CUDA
	* ``AXOM_USE_HIP`` - Defined if Axom is built with HIP

* ``AXOM_DEVICE_CODE`` - Denotes code will be compiled on device
* ``AXOM_GPUCC`` - Denotes code will be compiled with HIP or CUDA compiler

For ``__host__ __device__`` decoration::

  //---------------------------------------------------------------------------
  // Functions
  //---------------------------------------------------------------------------

  #define AXOM_DEVICE __device__

  #define AXOM_HOST_DEVICE __host__ __device__

  #define AXOM_HOST __host__

  //---------------------------------------------------------------------------
  // Lambdas
  //---------------------------------------------------------------------------

  #define AXOM_LAMBDA [=] AXOM_HOST_DEVICE

  #define AXOM_DEVICE_LAMBDA [=] AXOM_DEVICE

  #define AXOM_HOST_LAMBDA [=] AXOM_HOST

.. _gpu-memory-label:

===============
Memory 
===============

Link to Axom's `memory management header file`_

> (axom/core/memory_management.hpp)

Reviewer 2: This looks a bit funny in the rendered sphinx.

Would something like the following be clearer?
```
Axom's memory management functionality can be found in `axom/core/memory_management.hpp`_(<link>)
```

   :start-after:  _cuda_exec_start
   :end-before:  _cuda_exec_end
   :language: C++

Each execution space (``axom::execution_space<ExecSpace>``) provides:

* Axom policies that are type aliases of RAJA policies to be used with kernels,
  RAJA types, and RAJA operations

  * ``loop_policy`` - For `RAJA scans`_ and other operations; ``axom::for_all``
    uses the loop_policy from the templated execution space.
  * ``reduce_policy`` - For  `RAJA reduction types`_
  * ``atomic_policy`` - For `RAJA atomic operations`_
  * ``sync_policy`` - For Axom's `synchronize`_ function, which is a wrapper
    around ``RAJA::synchronize()``

* Umpire allocator defaults

  * ``memory_space`` - The memory space abstraction for use by
    :ref:`Core Containers<core-containers>` like ``axom::Array``.
  * ``allocatorID()`` - Gets the allocator ID for the Umpire resource to use in
    this execution space.

* General information on the execution space

  * ``name()`` - Name of the execution space
  * ``onDevice()`` - Is the execution space on device? (True/False)
  * ``valid()`` - Is the execution space valid? (True)
  * ``async()`` - Is the execution space asynchronous? (True/False)

> There is also a set of `nested execution policies`_
> (axom/mint/execution/internal/structured_exec.hpp) to be used with
> ``RAJA::kernel``, and are used primarily for iterating over mint meshes.

Reviewer 2: Suggestion/Clarification:

```suggestion
The mint component also provides a set of `nested execution policies`_
(axom/mint/execution/internal/structured_exec.hpp) to be used with
``RAJA::kernel`` e.g. for iterating over mint meshes.
```

* When you attempt to recompile your code, unless you have already done the
  ``__host__ __device__`` decoration correctly already, the compiler should
  complain about what functions are not yet compiling on device, and that you
  need to add ``__host__ __device__`` decorators::

    error: calling a __host__ function("XXX") from a __host__ __device__ function("YYY") is not allowed

  * Even after you add the decorators to the functions the compiler complains
    about, it is likely recompiling will introduce complaints about more
    functions (the functions being the non-decorated functions your
    newly-decorated functions are calling). You will want to keep decorating
    until all the complaints are gone.
  * Most of the C++ standard library is not available on device. Your options
    are Axom's equivalent functions/classes if it exists, or to add your
    own / rewrite the code to not use standard library.

* With no more decorating complaints from the compiler, write the logically
  correct kernel.

  * If at this point your kernel is not working/segfaulting, it's hopefully a
    logical error, and you can debug the kernel without diving into debugging
    tools.

>   * ``printf()`` for output

Reviewer 2: > * ``printf()`` for output

Please expand this line into a full sentence.

Also, should we mention which, if any, `slic` macros work in kernel code?

> .. ## Copyright (c) 2017-2022, Lawrence Livermore National Security,con LLC and

Reviewer 3: What is "Security,con LLC" ?

Reviewer 4: That's a typo. ',con' should be removed.

Reviewer 3: ```suggestion
.. ## Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and
```

.. ## other Axom Project Developers. See the top-level LICENSE file for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)

.. _gpu-porting-label:

*******************************
GPU Porting in Axom
*******************************

> Axom uses RAJA and Umpire as the main workhorses for GPU porting,

Reviewer 5: Should probably add a link to RAJA and Umpire here.

Maybe something like:

Axom uses the following two libraries as the main workhorses for GPU porting:

* RAJA (<- make this a link): handles software abstractions that enable architecture and programming model portability for HPC applications 
* Umpire (<-again link): resource management library that allows the discovery, provision, and management of memory on machines with multiple memory devices like NUMA and GPUs.

> with a set of convenience macros and `axom`-namespaced wrappers encapsulating

Reviewer 5: ```suggestion
with a set of convenience macros and ``axom``-namespaced wrappers encapsulating
```

Single backticks are italicized and double mark it as code

> with a set of convenience macros and `axom`-namespaced wrappers encapsulating
> commonly used RAJA and Umpire functions, and preset execution spaces for

Reviewer 3: In addition to Chris's point, I'd suggest a rephrase:
```suggestion
with a set of convenience macros and function wrappers in the ``axom`` namespace encapsulating
commonly-used RAJA and Umpire functions, and preset execution spaces for
```

host/device execution.

For the user's guide on using GPU utilities, see also

> :ref:`Core Acceleration<core-acceleration>`

Reviewer 5: ```suggestion
:ref:`Core Acceleration<core-acceleration>`.
```

.. _gpu-macros-label:

===============
Macros 
===============

Axom's macros can be found in file

> `axom/core/Macros.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/Macros.hpp>`_

Reviewer 5: ```suggestion
`axom/core/Macros.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/Macros.hpp>`_.
```

Most of the GPU-related macros are used to guard device code for compilation
time or for ``__host__ __device__`` decoration of functions/lambdas.

For guarding device code:

.. literalinclude:: ../../../axom/core/Macros.hpp
   :start-after:  _guarding_macros_start
   :end-before:  _guarding_macros_end
   :language: C++

> 

Reviewer 4: ```suggestion

.. note:: * Functions called in CUDA or HIP GPU device code require the ``__device__`` annotation.
              * Functions that will be called in device code *and* CPU host code require the ``__host__ __device__`` annotation.
```

> For ``__host__`` or `` __device__`` decoration, or both:

Reviewer 5: ```suggestion
For ``__host__`` or ``__device__`` decoration, or both:
```

For some reason the space was making this render wrong i think

> For ``__host__`` or `` __device__`` decoration, or both:

Reviewer 4: ```suggestion
The following code shows the macros used in Axom to apply these annotations.
```
Also, it may be a good idea to include a code example showing how these macros are used.

.. literalinclude:: ../../../axom/core/Macros.hpp
   :start-after:  _decorating_macros_start
   :end-before:  _decorating_macros_end
   :language: C++

.. _gpu-memory-label:

===============
Memory 
===============

Axom's memory management routines can be found in the file

> `axom/core/memory_management.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/memory_management.hpp>`_

Reviewer 5: ```suggestion
`axom/core/memory_management.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/memory_management.hpp>`_.
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Memory Management Routines
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Umpire has the concept of "allocators" associated with each
`memory resource type`_ (``umpire::resource::MemoryResourceType``)

To allocate memory on a particular resource, you use the ID for the allocator

> associated with the MemoryResourceType.

Reviewer 5: ```suggestion
associated with the ``MemoryResourceType``.
```

Reviewer 5: You can leave off the namespaces if you like it better but leave the backticks please.

You are able to set a default allocator, whereby all your memory allocations
will go on the resource associated with the allocator unless otherwise
specified::

>   //---------------------------------------------------------------------------

Reviewer 5: This is code highlighting as Python I think. Please change it so it is specifically C++.

  // Getters and Setters for Allocators and ID
  //---------------------------------------------------------------------------

  // Gets the allocator ID associated with the given MemoryResourceType
  int axom::getUmpireResourceAllocatorID(umpire::resource::MemoryResourceType resource_type)

  // Sets the default allocator associated with the given MemoryResourceType
  void axom::setDefaultAllocator(umpire::resource::MemoryResourceType resource_type)

  // Sets the default allocator associated with the given allocator ID
  void axom::setDefaultAllocator(int allocId)

  // Returns the current default allocator ID
  int axom::getDefaultAllocatorID()

  //---------------------------------------------------------------------------
  // Memory Allocation Functions
  //---------------------------------------------------------------------------

>   // Allocates n number of T elements in the memory space with the given

Reviewer 4: Better to use code file comments to include this code block rather than entering it directly in sphinx for consistency and maintenance.

  // allocator ID. If no ID provided, uses the current default
  T* axom::allocate(std::size_t n, int allocID = getDefaultAllocatorID())

  //  Deallocates the given pointer
  void axom::deallocate(T*& p)

  // Reallocates pointer to given size and with the given allocator ID.
  // If no ID provided, uses the current default.
  T* axom::reallocate(T* p, std::size_t n, int allocID = getDefaultAllocatorID())

  // Copies numbytes from src to dst
  axom::copy(void* dst, const void* src, std::size_t numbytes)

.. note::

  When Axom is built without Umpire, the getters and setters shown above
  become no-ops or are undefined, while the memory allocation functions
  default to C++ standard library functions with only allocation on the
  host (CPU):

  * ``axom::allocate`` calls ``std::malloc``
  * ``axom::deallocate`` calls ``std::free``
  * ``axom::reallocate`` calls ``std::realloc``
  * ``axom::copy`` calls ``std::memcpy``

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
MemorySpace
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. literalinclude:: ../../../axom/core/memory_management.hpp
   :start-after:  _memory_space_start
   :end-before:  _memory_space_end
   :language: C++

> These ``axom::MemorySpace`` enum values are used to define where in memory

Reviewer 4: ```suggestion
Axom provides the ``axom::MemorySpace`` enum type to define values indicating the memory space where data in
```

> ``axom::Array`` and ``axom::ArrayView`` types are defined.

Reviewer 4: ```suggestion
``axom::Array`` and ``axom::ArrayView`` lives.
```

``Dynamic`` allows you to define the location at runtime, with some caveats
(see :ref:`Core Containers<core-containers>` for more details and examples).

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Useful Links
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> `Umpire Tutorial`_ - 1st two sections cover Allocators and Resources

Reviewer 5: ```suggestion
`Umpire Tutorial`_ - First two sections cover ``Allocators`` and ``Resources``.
```

.. _gpu-kernels-label:

===============
Kernels
===============

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
axom::for_all
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> ``axom::for_all`` can be found in

Reviewer 4: ```suggestion
``axom::for_all`` can be found in the file
```

> `axom/core/execution/for_all.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/for_all.hpp>`_

Reviewer 5: ```suggestion
`axom/core/execution/for_all.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/for_all.hpp>`_.
```

> ``axom::for_all`` is a wrapper around `RAJA forall`_, a simple for loop.

Reviewer 4: ```suggestion
``axom::for_all`` is a wrapper around `RAJA forall`_, which is used to execute simple for-loop kernels.
```

> This is used the most in Axom, from unit tests to example drivers for loops

Reviewer 4: ```suggestion
This is used in Axom to execute for-loop style kernels that will be run on a GPU device, or on both a GPU device and a CPU host.
```

> on device::

Reviewer 4: ```suggestion
For example::
```

  template <typename ExecSpace, typename KernelType>
  void axom::for_all(const IndexType& N, KernelType&& kernel)

  template <typename ExecSpace, typename KernelType>
  void axom::for_all(const IndexType& begin, const IndexType& end, KernelType&& kernel)

.. note::

>   When Axom is built without RAJA, ``axom::for_all`` becomes a for loop on

Reviewer 5: ```suggestion
  When Axom is built without RAJA, ``axom::for_all`` becomes a ``for`` loop on
```

>   When Axom is built without RAJA, ``axom::for_all`` becomes a for loop on

Reviewer 4: ```suggestion
  When Axom is built without RAJA, ``axom::for_all`` becomes a for-loop on
```

  host (CPU).

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RAJA::kernel
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
``RAJA::kernel`` is for more complex and/or nested loops.

This is used infrequently, mainly seen only in a `few unit tests`_.

Your general go-to will be ``axom::for_all``.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Useful Links
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> `RAJA Loops`_ - Covers for_all, kernel, and other loops

Reviewer 5: ```suggestion
`RAJA Loops`_ - Covers ``for_all``, kernel, and other loops.
```

> `RAJA Loops`_ - Covers for_all, kernel, and other loops

Reviewer 4: ```suggestion
`RAJA Loops`_ - Covers ``RAJA::forall``, ``RAJA::kernel``, ``RAJA::launch`` kernel execution methods.
```

.. _gpu-execution-spaces-label:

================================
Execution Spaces & Policies
================================

Axom's execution spaces can be found in

> `axom/core/execution/execution_space.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/execution_space.hpp>`_

Reviewer 5: ```suggestion
`axom/core/execution/execution_space.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/execution_space.hpp>`_.
```

> Axom's execution spaces are derived from an execution_space traits class,

Reviewer 5: ```suggestion
Axom's execution spaces are derived from an ``execution_space`` traits class,
```

Reviewer 4: Also remove the comma after the word 'class'

containing RAJA execution policies and default Umpire memory allocators
associated with each space.

> Axom currently supports `4 execution spaces`_:

Reviewer 4: ```suggestion
Axom currently supports `four execution spaces`_:
```

* ``SEQ_EXEC`` - Sequential execution policies on host
* ``OMP_EXEC`` -  OpenMP execution policies on host
* ``CUDA_EXEC`` - CUDA execution policies in Unified Memory (host + device)
* ``HIP_EXEC`` - HIP execution policies in Unified Memory (host + device)

> Additionally, ``HIP_EXEC`` and ``CUDA_EXEC`` are templated by the

Reviewer 4: ```suggestion
Additionally, ``HIP_EXEC`` and ``CUDA_EXEC`` policies are templated by the
```

number of threads and SYNCHRONOUS or ASYNC execution.

.. literalinclude:: ../../../axom/core/execution/internal/cuda_exec.hpp
   :start-after:  _cuda_exec_start
   :end-before:  _cuda_exec_end
   :language: C++

> Each execution space (``axom::execution_space<ExecSpace>``) provides:

Reviewer 4: This seems disconnected from the list of execution spaces above. For example, is SEQ_EXEC used as a template parameter to axom::execution_space< >?

Also, what is SEQ_EXEC? A macro constant? A type? An enum value? That should be noted.

* Axom policies that are type aliases of RAJA policies to be used with kernels,
  RAJA types, and RAJA operations

  * ``loop_policy`` - For `RAJA scans`_ and other operations; ``axom::for_all``
    uses the loop_policy from the templated execution space.
  * ``reduce_policy`` - For  `RAJA reduction types`_
  * ``atomic_policy`` - For `RAJA atomic operations`_
  * ``sync_policy`` - For Axom's `synchronize`_ function, which is a wrapper

>     around ``RAJA::synchronize()``

Reviewer 5: A short description of what reduce, atomic, and sync is would be helpful.

Reviewer 4: Also a couple of simple examples of usage would be very helpful.

* Umpire allocator defaults

  * ``memory_space`` - The memory space abstraction for use by
    :ref:`Core Containers<core-containers>` like ``axom::Array``.
  * ``allocatorID()`` - Gets the allocator ID for the Umpire resource to use in
    this execution space.

* General information on the execution space

  * ``name()`` - Name of the execution space
  * ``onDevice()`` - Is the execution space on device? (True/False)
  * ``valid()`` - Is the execution space valid? (True)
  * ``async()`` - Is the execution space asynchronous? (True/False)

> The mint component also provides a set of nested execution policies

Reviewer 5: ```suggestion
The Mint component also provides a set of nested execution policies
```

Reviewer 5: Also maybe a link to mint?

Reviewer 4: Also, add a link to Mint docs would be good.

located at
`axom/mint/execution/internal/structured_exec.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/mint/execution/internal/structured_exec.hpp>`_
to be used with
``RAJA::kernel`` e.g. for iterating over mint meshes.

.. note::

>   When Axom is built without RAJA, only ``SEQ_EXEC`` is available

Reviewer 4: This note is a bit confusing. It's not clear what Umpire has to do with available execution policies since Umpire is not used to execute anything, only to manage memory.

  for host (CPU) execution. When Axom is built without Umpire, only
  ``SEQ_EXEC`` and ``OMP_EXEC`` is available for host (CPU) execution.

.. _gpu-tips-label:

================================
General, Rough Porting Tips
================================

> 

Reviewer 4: This whole section would be much clearer, if you walked through an example that illustrates what you are trying to say.

> * Start with figuring out what memory you need on device, and use Axom's Array

Reviewer 5: ```suggestion
* Start with figuring out what memory you need on device, and use Axom's ``Array``
```

> * Start with figuring out what memory you need on device, and use Axom's Array

Reviewer 4: ```suggestion
* Start with figuring out what memory you need on device, and use ``axom::Array``
```

For consistency with other usage on this page.

>   class and memory_managment functions to do the allocations.

Reviewer 5: ```suggestion
  class and ``memory_managment`` functions to do the allocations.
```

Reviewer 4: What does memory_management refer to here? Specifically, why the underscore between the words?

> * Using a barebones Axom's for_all kernel with a device policy, attempt to

Reviewer 5: ```suggestion
* Using a barebones ``axom::for_all`` kernel with a device policy, attempt to
```

Reviewer 4: What do you mean by 'barebones'?  If you use a colloquial term or phrase, it needs to be explained.

>   access/ manipulate the memory on device.

Reviewer 5: ```suggestion
  access and/or manipulate the memory on device.
```

>   * This does not have to be logically correct, you are just making sure you

Reviewer 4: This sentence is confusing.

    have done the allocation correctly (e.g. not segfaulting).

* Add the functions you want to call on device to the ``axom::for_all`` kernel.

>   * Does not have to be logically correct.

Reviewer 4: Not sure what this means.

* When you attempt to recompile your code, unless you have already done the
  ``__host__ __device__`` decoration correctly already, the compiler should

>   complain about what functions are not yet compiling on device, and that you

Reviewer 4: ```suggestion
  complain about what functions are not yet compiling for the device, and that you
```

  need to add ``__host__ __device__`` decorators::

>     error: calling a __host__ function("XXX") from a __host__ __device__ function("YYY") is not allowed

Reviewer 4: This error message does not correspond to what you are trying to explain here.

  * Even after you add the decorators to the functions the compiler complains
    about, it is likely recompiling will introduce complaints about more
    functions (the functions being the non-decorated functions your
    newly-decorated functions are calling). You will want to keep decorating
    until all the complaints are gone.
  * Most of the C++ standard library is not available on device. Your options
    are Axom's equivalent functions/classes if it exists, or to add your

>     own / rewrite the code to not use standard library.

Reviewer 5: ```suggestion
    own or rewrite the code to not use standard library.
```

* With no more decorating complaints from the compiler, write the logically
  correct kernel.

  * If at this point your kernel is not working/segfaulting, it's hopefully a
    logical error, and you can debug the kernel without diving into debugging
    tools.

>   * Utilize ``printf()`` for debugging output

Reviewer 5: Can you mention here which tools help debugging, totalview or whatever nvidia tools are available that i should probably know but don't.

.. ## Copyright (c) 2017-2022, Lawrence Livermore National Security,con LLC and
.. ## other Axom Project Developers. See the top-level LICENSE file for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)

.. _gpu-porting-label:

*******************************
GPU Porting in Axom
*******************************

Axom uses RAJA and Umpire as the main workhorses for GPU porting,

> with a set of convenience macros, `axom`-namespaced wrappers around commonly

Reviewer 4: ```suggestion
with a set of convenience macros and `axom`-namespaced wrappers encapsulating commonly
```

> used functions, and preset execution spaces for host/device execution.

Reviewer 4: ```suggestion
used RAJA and Umpire functions, and preset execution spaces for host/device execution.
```

For the user's guide on using GPU utilities, see also
:ref:`Core Acceleration<core-acceleration>`

.. _gpu-macros-label:

===============
Macros 
===============

> Axom's macros can be found in

Reviewer 4: ```suggestion
Axom's macros can be found in the file
```

`axom/core/Macros.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/Macros.hpp>`_

> Most of the GPU-related macros are used to guard device code before compilation

Reviewer 4: ```suggestion
Most of the GPU-related macros are used to guard device code for compilation
```

time or for ``__host__ __device__`` decoration of functions/lambdas.

For guarding device code:

* ``AXOM_USE_GPU`` - Defined if Axom is built with CUDA or HIP

	* ``AXOM_USE_CUDA`` - Defined if Axom is built with CUDA
	* ``AXOM_USE_HIP`` - Defined if Axom is built with HIP

* ``AXOM_DEVICE_CODE`` - Denotes code will be compiled on device
* ``AXOM_GPUCC`` - Denotes code will be compiled with HIP or CUDA compiler

> For ``__host__ __device__`` decoration::

Reviewer 4: ```suggestion
For ``__host__`` or `` __device__`` decoration, or both::
```

  //---------------------------------------------------------------------------
  // Functions
  //---------------------------------------------------------------------------

  #define AXOM_DEVICE __device__

  #define AXOM_HOST_DEVICE __host__ __device__

  #define AXOM_HOST __host__

  //---------------------------------------------------------------------------
  // Lambdas
  //---------------------------------------------------------------------------

  #define AXOM_LAMBDA [=] AXOM_HOST_DEVICE

  #define AXOM_DEVICE_LAMBDA [=] AXOM_DEVICE

  #define AXOM_HOST_LAMBDA [=] AXOM_HOST

.. _gpu-memory-label:

===============
Memory 
===============

> Axom's memory management functionality can be found in

Reviewer 4: ```suggestion
Axom's memory management routines can be found in the file
```

`axom/core/memory_management.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/memory_management.hpp>`_

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Memory Management Routines
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Umpire has the concept of "allocators" associated with each

> `memory resource type`_ (``umpire::resource::MemoryResourceType``)

Reviewer 4: ```suggestion
`Umpire memory resource type <https://github.com/LLNL/Umpire/blob/develop/src/umpire/resource/MemoryResourceTypes.hpp#L63>`_
```

Better link?

Reviewer 5: Please put a period at the end of this sentence as well

To allocate memory on a particular resource, you use the ID for the allocator
associated with the MemoryResourceType.

> You are able to set a default allocator, where all your memory allocations will

Reviewer 4: ```suggestion
You are able to set a default allocator, whereby all memory allocations will
```

go on the resource associated with the allocator unless otherwise specified::

  //---------------------------------------------------------------------------
  // Getters and Setters for Allocators and ID
  //---------------------------------------------------------------------------

>   // Gets the allocator ID associated with the given MemoryResourceType

Reviewer 4: It may be better and easier to maintain if code comments are used to document code blocks like this. That way the docs won't ned to change when the code does.

  // Sets the default allocator associated with the given MemoryResourceType
  void axom::setDefaultAllocator(umpire::resource::MemoryResourceType resource_type)

  // Sets the default allocator associated with the given allocator ID
  void axom::setDefaultAllocator(int allocId)

  // Returns the current default allocator ID
  int axom::getDefaultAllocatorID()

  //---------------------------------------------------------------------------
  // Memory Allocation Functions
  //---------------------------------------------------------------------------

  // Allocates n number of T elements in the memory space with the given
  // allocator ID. If no ID provided, uses the current default
  T* axom::allocate(std::size_t n, int allocID = getDefaultAllocatorID())

  //  Deallocates the given pointer
  void axom::deallocate(T*& p)

  // Reallocates pointer to given size and with the given allocator ID.
  // If no ID provided, uses the current default.
  T* axom::reallocate(T* p, std::size_t n, int allocID = getDefaultAllocatorID())

  // Copies numbytes from src to dst
  axom::copy(void* dst, const void* src, std::size_t numbytes)

.. note::

>   When Axom is built without Umpire, the getters and setters become no-ops or

Reviewer 4: ```suggestion
  When Axom is built without Umpire, the getters and setters shown above become no-ops or
```
